recipes/openai.py

```diff
@@ -0,0 +1,391 @@
+        self.openai_n = openai_n
+        self.openai_n_retries = openai_n_retries
+        self.response_parser = response_parser
+
+    def __call__(
+        self,
+        stream: Iterable[Dict],
+        *,
+        nlp: Language,
+        batch_size: int,
+        **kwargs,
+    ) -> Iterable[Dict]:
+        if self.segment:
+            stream = preprocess.split_sentences(nlp, stream)
+
+        stream = self.pipe(stream, nlp, batch_size, **kwargs)
+        stream = self.set_hashes(stream)
+        return stream
+
+    def pipe(self, stream: Iterable[Dict], nlp: Language, batch_size: int, **kwargs):
+        """Process the stream and add suggestions from OpenAI."""
+        stream = self.stream_suggestions(stream, batch_size)
+        stream = self.format_suggestions(stream, nlp=nlp)
+        return stream
+
+    def set_hashes(self, stream: Iterable[Dict]):
+        for example in stream:
+            yield set_hashes(example)
+
+    def update(self, examples: Iterable[Dict]) -> float:
+        """Update the examples that will be used in the prompt based on user flags."""
+        for eg in examples:
+            if PromptExample.is_flagged(eg):
+                self.add_example(PromptExample.from_prodigy(eg, self.label))
+        return 0.0
+
+    def add_example(self, example: PromptExample) -> None:
+        """Add an example for use in the prompts. Examples are pruned to the most recent max_examples."""
+        if self.max_examples:
+            self.examples.append(example)
+        if len(self.examples) >= self.max_examples:
+            self.examples = self.examples[-self.max_examples :]
+
+    def stream_suggestions(
+        self, stream: Iterable[Dict], batch_size: int
+    ) -> Iterable[Dict]:
+        """Get zero-shot or few-shot annotations from OpenAI.
+
+        Given a stream of input examples, we define a prompt, get a response from OpenAI,
+        and yield each example with their predictions to the output stream.
+        """
+        for batch in batch_sequence(stream, batch_size):
+            prompts = [
+                self._get_prompt(eg["text"], labels=self.labels, examples=self.examples)
+                for eg in batch
+            ]
+            responses = self._get_openai_response(prompts)
+            for eg, prompt, response in zip(batch, prompts, responses):
+                if self.verbose:
+                    rich.print(Panel(prompt, title="Prompt to OpenAI"))
+                eg["openai"] = {"prompt": prompt, "response": response}
+                if self.verbose:
+                    rich.print(Panel(response, title="Response from OpenAI"))
+                yield eg
+
+    def format_suggestions(
+        self, stream: Iterable[Dict], *, nlp: Language
+    ) -> Iterable[Dict]:
+        """Parse the examples in the stream and set up labels
+        to display in the Prodigy UI.
+        """
+        stream = prodigy.components.preprocess.add_tokens(nlp, stream, skip=True)  # type: ignore
```

Comment:
Nitpick, but we already import `preprocess`, so we should be able to just: 

```suggestion
        stream = preprocess.add_tokens(nlp, stream, skip=True)  # type: ignore
```